packages/core/src/utils/SchemaQueries.ts

```diff
@@ -0,0 +1,73 @@
+import { type GeneratedSchema } from '@logto/schemas';
+import { type UpdateWhereData, type OmitAutoSetFields, type SchemaLike } from '@logto/shared';
+import { type CommonQueryMethods } from 'slonik';
+
+import { buildDeleteByIdWithPool } from '#src/database/delete-by-id.js';
+import { buildFindAllEntitiesWithPool } from '#src/database/find-all-entities.js';
+import { buildFindEntityByIdWithPool } from '#src/database/find-entity-by-id.js';
+import { buildInsertIntoWithPool } from '#src/database/insert-into.js';
+import { buildGetTotalRowCountWithPool } from '#src/database/row-count.js';
+import { buildUpdateWhereWithPool } from '#src/database/update-where.js';
+
+/**
+ * Query class that contains all the necessary CRUD queries for a schema. It is
+ * designed to be used with SchemaActions for a SchemaRouter. You can also extend
+ * this class to add more queries.
+ */
+export default class SchemaQueries<
+  Key extends string,
+  CreateSchema extends Partial<SchemaLike<Key> & { id: string }>,
+  Schema extends SchemaLike<Key> & { id: string },
+> {
+  #findTotalNumber: () => Promise<{ count: number }>;
+  #findAll: (limit: number, offset: number) => Promise<readonly Schema[]>;
+  #findById: (id: string) => Promise<Readonly<Schema>>;
+  #insert: (data: OmitAutoSetFields<CreateSchema>) => Promise<Readonly<Schema>>;
+
+  #updateById: <SetKey extends Key | 'id', WhereKey extends Key | 'id'>(
+    data: UpdateWhereData<SetKey, WhereKey>
+  ) => Promise<Schema>;
+
+  #deleteById: (id: string) => Promise<void>;
+
+  constructor(
+    public readonly pool: CommonQueryMethods,
+    public readonly schema: GeneratedSchema<Key | 'id', CreateSchema, Schema>
+  ) {
+    this.#findTotalNumber = buildGetTotalRowCountWithPool(this.pool, this.schema.table);
+    this.#findAll = buildFindAllEntitiesWithPool(this.pool)(this.schema);
+    this.#findById = buildFindEntityByIdWithPool(this.pool)(this.schema);
+    this.#insert = buildInsertIntoWithPool(this.pool)(this.schema, { returning: true });
+    this.#updateById = buildUpdateWhereWithPool(this.pool)(this.schema, true);
+    this.#deleteById = buildDeleteByIdWithPool(this.pool, this.schema.table);
+  }
+
+  async findTotalNumber(): Promise<number> {
+    const { count } = await this.#findTotalNumber();
+    return count;
+  }
+
+  async findAll(limit: number, offset: number): Promise<readonly Schema[]> {
+    return this.#findAll(limit, offset);
```

Comment:
nit: feels like might be an over-design here.  Could simply keep the public copy of all the query methods? 

Most likely the subclass will overwrite the implementation of a  query method entirely without referencing the original built one?

Reply:
didn't get this one. let's discuss offline

Reply:
discussed offline

Reply:
misunderstood before